### PR TITLE
os/bluestore: memorize layout of BlueFS on management.

### PR DESCRIPTION
**Pull request by rzarzynski**
Fixes: https://tracker.ceph.com/issues/41744
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -6541,7 +6550,14 @@ int BlueStore::migrate_to_new_bluefs_device(const set<int>& devs_source,
   bluefs->add_block_extent(
     id, reserved, bluefs->get_block_device_size(id) - reserved);
 
-  r = bluefs->device_migrate_to_new(cct, devs_source, id);
+  if (!link_db.empty() && id != BlueFS::BDEV_NEWDB) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: This part probably lacks db creation case, i.e. when link_db.empty() && id == NEWDB.
Then we starts having dedicated db device...

**Comment by tchaikov** [comment]
```
2019-09-28T15:50:58.340 INFO:tasks.workunit.client.0.smithi014.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_S
IZE/huge/release/15.0.0-5643-ga29a4f4/rpm/el7/BUILD/ceph-15.0.0-5643-ga29a4f4/src/os/bluestore/BlueStore.cc: 5380: FAILED ceph_assert(bluefs->maybe_verify_layout(bluefs_layout) == 0)
2019-09-28T15:50:58.340 INFO:tasks.workunit.client.0.smithi014.stderr:
2019-09-28T15:50:58.340 INFO:tasks.workunit.client.0.smithi014.stderr: ceph version 15.0.0-5643-ga29a4f4 (a29a4f453b5179db85776998a5fb826f35aaf732) octopus (dev)
2019-09-28T15:50:58.340 INFO:tasks.workunit.client.0.smithi014.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x14a) [0x7fe6fc132a2b]
2019-09-28T15:50:58.340 INFO:tasks.workunit.client.0.smithi014.stderr: 2: (()+0x26abf3) [0x7fe6fc132bf3]
2019-09-28T15:50:58.341 INFO:tasks.workunit.client.0.smithi014.stderr: 3: (BlueStore::_open_bluefs(bool)+0x208) [0x55a308bc60e8]
2019-09-28T15:50:58.341 INFO:tasks.workunit.client.0.smithi014.stderr: 4: (BlueStore::_mount_for_bluefs()+0x41) [0x55a308bc6141]
2019-09-28T15:50:58.341 INFO:tasks.workunit.client.0.smithi014.stderr: 5: (BlueStore::migrate_to_existing_bluefs_device(std::set<int, std::less<int>, std::allocator<int> > const&, int)+0x77) [0x55a308c244
27]
```

http://pulpito.ceph.com/kchai-2019-09-28_11:33:27-rados-wip-kefu-testing-2019-09-28-1615-distro-basic-smithi/4342614/

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (rzarzynski) [reply src/os/bluestore/BlueStore.cc]: Reworked.

**Comment by rzarzynski** [comment]
Pushed updated version. Teuthology job: http://pulpito.front.sepia.ceph.com/rzarzynski-2019-10-03_11:54:28-rados-wip-bs-memorize-layout-on-management-distro-basic-smithi/4353921/.

**Pull request merged**